cvat/apps/engine/static/engine/js/player.js

```diff
@@ -14,121 +14,228 @@
     Mousetrap:false
 */
 
-'use strict';
-
-class FrameProvider extends Listener {
-    constructor(stop, tid) {
+class FrameProviderWrapper extends Listener {
+    constructor(stop) {
         super('onFrameLoad', () => this._loaded);
-        this._MAX_LOAD = 500;
 
-        this._stack = [];
-        this._loadInterval = null;
-        this._required = null;
-        this._loaded = null;
-        this._loadAllowed = true;
-        this._preloadRunned = false;
-        this._loadCounter = this._MAX_LOAD;
-        this._frameCollection = {};
         this._stop = stop;
-        this._tid = tid;
+        this._loaded = null;
+        this._result = null;
+        this._required = null;
     }
 
-    require(frame) {
-        if (frame in this._frameCollection) {
-            this._preload(frame);
-            return this._frameCollection[frame];
+    async require(frameNumber) {
+        this._required = frameNumber;
+        const frameData = await window.cvatTask.frames.get(frameNumber);
+        const ranges = await window.cvatTask.frames.ranges();
+        for (const range of ranges) {
+            const [start, stop] = range.split(':').map((el) => +el);
+            if (frameNumber >= start && frameNumber <= stop) {
+                const data = await frameData.data();
+                this._loaded = frameNumber;
+                this._result = {
+                    data,
+                    renderWidth: frameData.width,
+                    renderHeight: frameData.height,
+                };
+                return this._result;
+            }
         }
-        this._required = frame;
-        this._loadCounter = this._MAX_LOAD;
-        this._load();
+
+        // fetching from server
+        // we don't want to wait it
+        // but we promise to notify the player when frame is loaded
+        frameData.data().then((data) => {
+            this._loaded = frameNumber;
+            this._result = {
+                data,
+                renderWidth: frameData.width,
+                renderHeight: frameData.height,
+            };
+
+            this.notify();
+        }).catch(() => {
+            this._loaded = { frameNumber };
+            this.notify();
+        });
+
         return null;
     }
 
-    _onImageLoad(image, frame) {
-        const next = frame + 1;
-        if (next <= this._stop && this._loadCounter > 0) {
-            this._stack.push(next);
-        }
-        this._loadCounter--;
-        this._loaded = frame;
-        this._frameCollection[frame] = image;
-        this._loadAllowed = true;
-        image.onload = null;
-        image.onerror = null;
-        this.notify();
+    get loaded() {
+        return this._loaded;
     }
 
-    _preload(frame) {
-        if (this._preloadRunned) {
-            return;
+    get result() {
+        return this._result;
+    }
+}
+
+class FrameBuffer extends Listener {
```

Comment:
Why have you implemented this non trivial logic outside of cvat-core? These old files will be removed soon.